### PR TITLE
Updated plugin to v3.

### PR DESCRIPTION
**Pull request by Choco31415**

### CODE REVIEW

File: ScratchBlocks.php [hunk 1]
```diff
@@ -7,8 +7,8 @@
  * MIT Licensed
  * http://opensource.org/licenses/MIT
  *
- * Includes scratchblocks2
- * http://github.com/blob8108/scratchblocks2
+ * Includes scratchblocks v3
+ * ??? Please get github link and put here
```
Comment (tjvr) [inline ScratchBlocks.php]: Please could you do this? :)

File: ScratchBlocks.php [hunk 2]
```diff
@@ -61,17 +52,18 @@ function sbSetup () {
...
-    'styles' => 'scratchblocks2/build/scratchblocks2.css',
+    'styles' => 'ScratchBlocks/src/defs.css',
 
     // jQuery is loaded anyway
     'dependencies' => array(),
 
     // Where the files are
     'localBasePath' => __DIR__,
-    'remoteExtPath' => 'mw-ScratchBlocks2'
+    'remoteExtPath' => 'ScratchBlocks'
```
Comment (tjvr) [inline ScratchBlocks.php]: Should this be `wiki-scratchblocks`?

**Comment by Choco31415** [comment]
Truthfully, I'm not 100% sure what 'remoteExtPath' is. It does't seem to be used by anything...

I'll fix the comment.

**Comment by tjvr** [comment]
Please can we keep the submodules, and not change scratchblocks.js? This would make the plugin easier to update. :)

**Comment by Choco31415** [comment]
If you show me how to render multiple (2+) languages at a time, then maybe. ;)

**Comment by tjvr** [comment]
```
renderMatching("pre.blocks", {
  languages: ['en', 'de', 'fr'],
})
```

**Comment by Choco31415** [comment]
It should be possible to replace the inner Scratchblocks folder with a redirect to the scratchblocks repo now.

**Comment by tjvr** [comment]
@Choco31415 Maybe try reverting the commit where you removed gitmodules? :)

**Comment by Choco31415** [comment]
The repo already had the original gitmodules file. It doesn't seem to be doing anything.

**Comment by tjvr** [comment]
It contains the submodules...

**Comment by Choco31415** [comment]
The submodule is added.

**Comment by tjvr** [comment]
I'm impressed (and confused!) as to why you deleted every file one-by-one! :) Not that it should matter.

File: ScratchBlocks.php [hunk 3]
```diff
@@ -22,33 +22,24 @@
...
+// Called to ouput HTML for <scratchsig> tag
 
 function sbRenderTag ($input, array $args, Parser $parser, PPFrame $frame) {
     return '<pre class="blocks">' . htmlspecialchars($input) . '</pre>';
 }
 
 
-// Output HTML for <scratchblocks> tag
-
-function sbRenderInlineTag ($input, array $args, Parser $parser,
```
Comment (tjvr) [inline ScratchBlocks.php]: Why did you remove the inline tag? :(

File: ScratchBlocks.php [hunk 4]
```diff
@@ -22,33 +22,24 @@
...
     // Register <scratchblocks> tag
     $parser->setHook('scratchblocks', 'sbRenderTag');
-    $parser->setHook('sb', 'sbRenderInlineTag');
     return true;
 }
+ 
 
-
-// Output HTML for <scratchblocks> tag
+// Called to ouput HTML for <scratchsig> tag
```
Comment (tjvr) [inline ScratchBlocks.php]: This is not `scratchsig` :-)

File: ScratchBlocks.php [hunk 5]
```diff
@@ -61,17 +52,18 @@ function sbSetup () {
...
     'scripts' => array(
-        'scratchblocks2/build/scratchblocks2.js',
-        'run_scratchblocks2.js',
+        'ScratchBlocks/src/scratchblocks.js',
+		'scratchblocks/src/translations.js',
+        'run_scratchblocks.js',
     ),
 
-    'styles' => 'scratchblocks2/build/scratchblocks2.css',
+    'styles' => 'ScratchBlocks/src/defs.css',
```
Comment (tjvr) [inline ScratchBlocks.php]: We don't want this here. These definitions are included in the JS (and then in the generated SVGs). :)

**Comment by Choco31415** [comment]
It was a good learning experience. :p I probably did it because I didn't want multiple repos over multiple computers.

I never noticed the removal of the inline Scratchblocks. I will add and fix those two issues.

**Comment by Choco31415** [comment]
Updated, and everything is confirmed as working.

**Comment by tjvr** [comment]
Thanks! Have you checked the inline tag `<sb>` works?

**Comment by Choco31415** [comment]
Yes, however it looks like something is preventing it from being inline:

http://i.cubeupload.com/P88RU8.png

I checked, and the code block has the "inline" style.

**Comment by tjvr** [comment]
Just add a CSS file that targets `code.blocks` or something :-)

**Comment by tjvr** [comment]
Alternatively: how are those blocks being rendered? You haven't called `scratchblocks.renderMatching("code.blocks")`, which is what I would expect.

I would have expected you need to add the following line to `run_scratchblocks.js`:

```diff
+ scratchblocks.renderMatching("code.blocks", {
+   inline: true,
+   languages: ["en", "de", "es", "el", "nl", "pt", "zh_CN", "tr", "nb", "ko", "it", "id", "ru", "ca", "ja", "pl", "he", "fr"],
+ });
```

**Comment by Choco31415** [comment]
I am calling that, it's just not pushed to the repo yet.

**Comment by labster** [comment]
Hi guys, I was going to suggest a PR that adds a credits section on Special:Version, and maybe support for wfLoadExtension() ... but this PR is scary.  Just commenting here so that I can keep track of when this gets merged.

By the way, I'm one of the technical staff at @miraheze, a free nonprofit wiki farm.  I have no idea what Scratch even is, but one of our wikis asked to install it this extension.  It looks good so far, but looks like I'll have to do security review again when the PR is ready.  Well :+1: for progress.

**Comment by Choco31415** [comment]
I know the PR is a bit of a mess... I'm honestly a bit new to this stuff, both MW extension building and forks. However, I'm learning and always open to requests and suggestions.

Since you commented, I will try and add a credits section and wfLoadExtension support, as requested. I will also try to consolidate the number of commits I'm making to keep things less "scary".

**Comment by enzanki-ars** [comment]
@tjvr Before I help @Choco31415, I just want to make sure... Would rebasing this PR be an acceptable action to take?

**Comment by tjvr** [comment]
Yup! rebasing to squash the extra commits would be an excellent idea. :-)

**Comment by Choco31415** [comment]
I'm doing that right now.

**Comment by Choco31415** [comment]
Okay, I made a version of this plugin that uses wfLoadExtension instead of require_once:

https://github.com/Choco31415/ScratchBlocks

@blob8108:
You're going to need to maintain two separate wiki plugin repos. The wfLoadExtension version only works with MW 1.25+, while the require_once version has been deprecated 1.25+. It will be important to distinguish between the two.

**Comment by Choco31415** [comment]
Edit: I am currently looking into how to best handle the wfLoadExtension support and the require_once support.

**Comment by labster** [comment]
While wfLoadExtension has deprecated replaced require_once loading a year ago, the documentation on extension.json is abysmal.  Because of this and other factors, at least half of the ecosystem hasn't moved to the new system.  Note: Even Wikia is still on 1.19.24, so it's not like Mediawiki installs are kept up to date.

As I said in the other ticket, branches are superior to other repositories -- and integrating the code into the same branch is even better.

File: ScratchBlocks.php [hunk 6]
```diff
@@ -74,3 +74,12 @@ function sbSetup () {
...
     'remoteExtPath' => 'ScratchBlocks'
 );
+
+$wgExtensionCredits['parserhook'][] = array(
+    'name' => "Scratchblocks",           // Name of extension - string
+    'description' => "This plugin takes text-based Scratch code and renders it in a graphical format.",    // Description of what the extension does - string. Omit in favour of descriptionmsg.
+    'version' => 3.1,         // Version number of extension - number or string
+    'author' => ["ErnieParke","blob8108"],         // The extension author's name - string or array for multiple
+    'url' => "https://github.com/tjvr/wiki-scratchblocks",            // URL of extension (usually instructions) - string
+    'license-name' => "https://github.com/tjvr/wiki-scratchblocks/blob/master/LICENSE",   // Short name of the license, links LICENSE or COPYING file if existing - string, added in 1.23.0
```
Comment (labster) [inline ScratchBlocks.php]: `'license-name'` should be `'MIT'`

**Comment by Choco31415** [comment]
The LICENSE is made by blob8108, not MIT. I'm also 90% sure that the plugin was made by blob8108. As such, the plugin shouldn't be licensed to MIT.

**Comment by labster** [comment]
I can clarify with this: https://www.mediawiki.org/wiki/Manual:$wgExtensionCredits#license-name  It's a technical limitation.  The name of the open source license is MIT.  That does not mean the code is licensed to MIT.

**Comment by Choco31415** [comment]
Huh, that is quite the coincidence. I will push that then.

**Comment by tjvr** [comment]
We can create two separate branches if we must, but I'd much rather have it all in one branch, and it sounds like that's possible.

Let me know when you've figured all this out and you're happy.

**Comment by Choco31415** [comment]
I did more investigation into this issue, and it actually does seem safe and acceptable to support both wfLoadExtension and require_once in the same extension. I will merge the branches later today.

**Comment by Choco31415** [comment]
The plugin can now use both require_once and wfLoadExtension. Anything else blob8108 and labster?

File: ScratchblockHooks.php [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+	
+	public static function sbRenderTag ($input, array $args, Parser $parser, PPFrame $frame) {
+		//throw new Exception("234");
+		return '<pre class="blocks">' . htmlspecialchars($input) . '</pre>';
+	}
+
+	// Output HTML for inline <sb> tag
+	public static function sbRenderInlineTag ($input, array $args, Parser $parser, PPFrame $frame) {
+		//throw new Exception("what");
+		//throw new Exception("2345");
```
Comment (labster) [inline ScratchblockHooks.php]: throw new Exception("testing code left in main 😄 ");

**Comment by labster** [comment]
Actually give me a few minutes, I'm going to PR your PR @Choco31415.

**Comment by Choco31415** [comment]
I have re-tested the code with the merge, and it works properly.

**Comment by banana439monkey** [comment]
Will the forum plugin have angled booleans?

**Comment by Choco31415** [comment]
I don't see a reason why it would not, assuming Scratch updates the forum Scratchblocks plugin.

File: .gitattributes [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+*.doc	 diff=astextplain
+*.DOC	 diff=astextplain
+*.docx diff=astextplain
+*.DOCX diff=astextplain
+*.dot  diff=astextplain
+*.DOT  diff=astextplain
+*.pdf  diff=astextplain
+*.PDF	 diff=astextplain
+*.rtf	 diff=astextplain
+*.RTF	 diff=astextplain
```
Comment (tjvr) [inline .gitattributes]: Please remove this file :)

File: run_scratchblocks.js [hunk 9]
```diff
@@ -0,0 +1,2 @@
+scratchblocks.renderMatching("pre.blocks", {languages: ["en", "de", "es", "el", "nl", "pt", "zh_CN", "tr", "nb", "ko", "it", "id", "ru", "ca", "ja", "pl", "he", "fr"],});
+scratchblocks.renderMatching("code.blocks", {languages: ["en", "de", "es", "el", "nl", "pt", "zh_CN", "tr", "nb", "ko", "it", "id", "ru", "ca", "ja", "pl", "he", "fr"],});
```
Comment (tjvr) [inline run_scratchblocks.js]: Should this line include `inline: true`?

**Comment by tjvr** [comment]
@Choco31415 Excellent work! I just have a couple more comments (see above), and then assuming @labster's happy I'll merge this. :)

**Comment by Choco31415** [comment]
I will remove it later. As for "inline: true", where should that be?

File: .gitattributes (hunk 8 above)
Comment (labster) [inline .gitattributes]: I'd suggest keeping the file, but keeping only lines 1 and 2.  It's useful for getting contributions from Windows folk who accidentally like to keep putting ^M in my vim.  The other types are not likely to ever occur in this repo.

**Comment by labster** [comment]
I think he's saying that inline:true come after the languages option.  In particular, that option does this before parsing:
```
    if (options.inline) {
      code = code.replace(/\n/g, ' ');
    }
```
Which I think ensures it only occupies one line?  So you should either add it to the code.blocks one, or alternatively just do one call to renderMatching, selecting on `"pre.blocks, code.blocks"`.

**Comment by Choco31415** [comment]
<sb> tags are already ensured to occupy one line because of the inline.css file. Besides, removing \n before an <sb> tag will not make it inline. Changing the <sb> display type will.

**Comment by labster** [comment]
Agreed.  Setting `inline: true` only makes rendering worse, as it will only let you make one block, as far as I can tell.  I'm not seeing any advantages to forcing that on an editor.

Proposing some minor changes in Choco31415/wiki-scratchblocks#2

But after that, everything looks good to me 👍

**Comment by Choco31415** [comment]
Everything does not look good. There seems to be a lag issue.

Earlier I got to test the Scratchblocks v3 plugin on the international Test Wiki. MartinWollenweber and I tried it on this page: http://test.scratch-wiki.info/wiki/User:Mtwoll/s2

The page currently uses v2 of the Scratchblocks plugin, and the page renders all tags in less then a second. However, when I was testing v3, the page was taking over a minute to fully render. See:

http://i.cubeupload.com/Nie9NC.png
http://i.cubeupload.com/jItp3L.png

Basically, the plugin would render one tag a second.

**Comment by labster** [comment]
There's nothing we can do about the lag issue in this repo, unless we need to pass the `go-faster` option :D.  I encourage you to open an upstream issue.

**Comment by Choco31415** [comment]
That issue is fixed.

**Comment by labster** [comment]
@Choco31415 You'll need to bump the commit for the scratchblocks submodule to the current master if you want the speed fixes.

**Comment by tjvr** [comment]
@Choco31415 This looks great.

Thank you so much for doing this!

Do let me know if any of the wikis have a problem with it. :)
